python/kserve/kserve/protocol/rest/openai/openai_proxy_model.py

```diff
@@ -0,0 +1,230 @@
+        This method will be called once for each chunk that is streamed back to the user.
+        """
+        pass
+
+    def _handle_completion_chunk(self, raw_chunk: str, request: CompletionRequest):
+        # Skip empty lines
+        if len(raw_chunk) == 0:
+            return None
+        # All chunks are prefixed with "data: "
+        data = raw_chunk[6:]
+        if data == "[DONE]":
+            return None
+
+        if self.skip_upstream_validation:
+            obj = orjson.loads(data)
+            completion_chunk = Completion.model_construct(**obj)
+        else:
+            completion_chunk = Completion.model_validate_json(data)
+        self.postprocess_completion_chunk(completion_chunk, request)
+        return completion_chunk
+
+    def _handle_chat_completion_chunk(
+        self, raw_chunk: str, request: ChatCompletionRequest
+    ):
+        # Skip empty lines
+        if len(raw_chunk) == 0:
+            return None
+        # All chunks are prefixed with "data: "
+        if len(raw_chunk) == 0:
+            return None
+        data = raw_chunk[6:]
+        if data == "[DONE]":
+            return None
+        if self.skip_upstream_validation:
+            obj = orjson.loads(data)
+            chat_completion_chunk = ChatCompletionChunk.model_construct(**obj)
+        else:
+            chat_completion_chunk = ChatCompletionChunk.model_validate_json(data)
+        self.postprocess_chat_completion_chunk(chat_completion_chunk, request)
+        return chat_completion_chunk
+
+    async def _stream(self, endpoint: str, content: str) -> httpx.Response:
+
+        req = self._http_client.build_request(
+            "POST",
+            endpoint,
+            content=content,
+            headers={"Content-type": "application/json"},
+        )
+        r = await self._http_client.send(req, stream=True)
+        return r
+
+    async def create_completion(
+        self, request: CompletionRequest
+    ) -> Union[Completion, AsyncIterator[Completion]]:
+        self.preprocess_completion_request(request)
+        if request.params.stream:
+            r = await self._stream(
+                self._completions_endpoint,
+                request.params.model_dump_json(exclude_none=True, exclude_unset=True),
+            )
+            it = AsyncMappingIterator(
+                iterator=r.aiter_lines(),
+                mapper=partial(self._handle_completion_chunk, request=request),
+                close=r.aclose,
+            )
+            return it
+        else:
+            response = await self._http_client.post(
+                self._completions_endpoint,
+                content=request.params.model_dump_json(
+                    exclude_none=True, exclude_unset=True
+                ),
+                headers={"Content-type": "application/json"},
+            )
+            if self.skip_upstream_validation:
+                obj = response.json()
```

Comment:
If we receive a 404 or any non 2xx or a non json body, this code will throw. Client will see a 500 with a cryptic error. It is nice to have better error handling. Some errors we can forward back to the clients, some of them we can give a better messaging etc.

Reply:
I agree error handling is woefully inadequate at this point. I'm working on adding in proper error handling of upstream issues (i.e. non-200, server offline, etc.)